### PR TITLE
Globally improve android frontend dev environment

### PR DESCRIPTION
**Pull request by GabrielFleischer**
This PR aims at improving all these little issues that appear around the actual project :

- [x] Bump Gradle version
- [x] Fix JaCoCo errors when running tests (achieved by updating Hilt)
- [x] Rework the layout of the test directories
- [x] Improve CI readability
- [x] Improve logs of tests (I would love to add colors, but I will try later)
- [x] Fix failing tests
- [x] Upgrade all dependencies and make sure all are useful
- [x] Fix codesmells
- [x] Review coverage exclusion policy to make sure auto-generated classes are excluded
- [x] Create a test template file that can be used as a basis of any UI test
- [x] Update documentation to match the new state, especially regarding testing

The new test layout is as follows :
```
src/
├─ main/              -- contains the code of the app
├─ test/              -- everything needed for the tests
│  ├─ framework/      -- contains the utility code that are used by the tests
│  │  ├─ common/      -- the classes used in common by both test types
│  │  ├─ emulator/    -- the classes used only by the emulated tests
│  │  ├─ robolectric/ -- the classes used only by the robolectric tests
│  ├─ ui/             -- contains the ui tests
│  │  ├─ emulator/    -- contains the tests run on the emulator
│  │  ├─ robolectric/ -- contains the tests that will be run in robolectric
│  ├─ unit/           -- contains the units tests
```

Its main goal is to allow developers to quickly run a test in the CI without too much work. 

I also added a Gradle plugin to improve the test output, now, it looks like this :
image

### CODE REVIEW

File: fe2-android/app/src/test/framework/emulator/java/com/github/dedis/popstellar/HiltTestRunner.java [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+
+/**
+ * JUnit Runner provider the HiltTestApplication that will inject dependencies into the tested code.
+ *
+ * <p>All tests using hilt to inject dependencies should be annotated with {@link HiltAndroidTest}
+ *
+ * <p>For more information : <a
+ * href="https://developer.android.com/training/dependency-injection/hilt-testing">https://developer.android.com/training/dependency-injection/hilt-testing</a>
+ */
+@SuppressWarnings("unused")
```
Comment (jbsv) [inline fe2-android/app/src/test/framework/emulator/java/com/github/dedis/popstellar/HiltTestRunner.java:19]: suppressing a warning generally deserves a comment ;-)

File: fe2-android/docs/README.md [hunk 2]
```diff
@@ -335,6 +344,29 @@ To inject custom modules using Hilt,
...
+will be executed on the emulator and will give a visual feedback. And all tests in `src/test/ui/robolectric`
+will be executed using the framework.
+
+Beware, only the robolectric tests are executed on the CI. You need to place your tests in the
+robolectric directory once you finished writing it.
+
+### Getting Started
+
+A template UI test has been written in the emulator directory. It is advised to take a look at it
+and use it as a basis when writing tests.used
```
Comment (jbsv) [inline fe2-android/docs/README.md:366]: `tests.used` ?

**Review by jbsv — DISMISSED** [review]
Thanks for this huge work. A few small comments.

File: fe2-android/app/build.gradle [hunk 3]
```diff
@@ -49,38 +56,77 @@ android {
...
     }
 
-    // The release tests cannot work as they need the debug-only activity EmptyActivity
-    // This configuration is only there to keep the usual test folder instead of a testDebug folder
     sourceSets {
+        // Define the project's file layout. For more information, please read the related documentation
+        //
+        // But, if you need to know the exact behavior, here it is :
+        // The idea is that we use a modified debug sourceset to hold the shared sources between the tests.
+        // This means that the debug version of the app will hold the test sources.
```
[the remark this answers is not in the document]
Comment (GabrielFleischer) [reply fe2-android/app/build.gradle:89]: In the industry, I would do something that is less intuitive to a new student.

Also, it does not add anything to the release version of the app. And there is no entry

If you want to, I can create a new build variant other than debug and put the tests there.

File: fe2-android/app/src/test/framework/emulator/java/com/github/dedis/popstellar/HiltTestRunner.java (hunk 1 above)
Comment (GabrielFleischer) [reply fe2-android/app/src/test/framework/emulator/java/com/github/dedis/popstellar/HiltTestRunner.java:19]: Ok

File: fe2-android/docs/README.md (hunk 2 above)
Comment (GabrielFleischer) [reply fe2-android/docs/README.md:366]: I... don't know 😅

File: fe2-android/app/build.gradle (hunk 3 above)
[the remark this answers is not in the document]
Comment (jbsv) [reply fe2-android/app/build.gradle:89]: Yes, it is very common to release debug versions to a subset of customers or partners who agree to adopt an early version and help testing in the field.

**Review by jbsv — APPROVED** [review]

**Pull request merged**